core/shared/src/main/scala/scalaz/zio/Managed.scala

```diff
@@ -11,44 +11,36 @@ package scalaz.zio
  * has been consumed, the resource will not be valid anymore and may fail with
  * some checked error, as per the type of the functions provided by the resource.
  */
-sealed abstract class Managed[+E, +R] extends Serializable { self =>
-  type R0 <: R
+final case class Managed[+E, +R](setup: IO[E, Managed.UseToken[E, R]]) { self =>
+  import Managed.UseToken
 
-  protected def acquire: IO[E, R0]
-
-  protected def release: R0 => IO[Nothing, _]
-
-  def use[E1 >: E, A](f: R => IO[E1, A]): IO[E1, A]
+  def use[E1 >: E, A](f: R => IO[E1, A]): IO[E1, A] =
+    setup.flatMap { token =>
+      token.acquire.flatMap(f).ensuring(token.release)
+    }
 
   final def use_[E1 >: E, A](f: IO[E1, A]): IO[E1, A] =
     use(_ => f)
 
   final def map[R1](f0: R => R1): Managed[E, R1] =
-    new Managed[E, R1] {
-      type R0 = R1
-
-      protected def acquire: IO[E, R1] = IO.never
-
-      protected def release: R1 => IO[Nothing, Unit] = _ => IO.unit
-
-      final def use[E1 >: E, A](f: R1 => IO[E1, A]): IO[E1, A] =
-        self.use(r => f(f0(r)))
+    Managed[E, R1] {
+      self.setup.map(token => token.copy(acquire = token.acquire.map(f0)))
     }
 
   final def flatMap[E1 >: E, R1](f0: R => Managed[E1, R1]): Managed[E1, R1] =
-    new Managed[E1, R1] {
-      type R0 = R1
-
-      protected def acquire: IO[E1, R1] = IO.never
-
-      protected def release: R1 => IO[Nothing, Unit] = _ => IO.unit
-
-      final def use[E2 >: E1, A](f: R1 => IO[E2, A]): IO[E2, A] =
-        self.use { r =>
-          f0(r).use { r1 =>
-            f(r1)
-          }
-        }
+    Managed[E1, R1] {
+      Ref.make[List[IO[Nothing, Any]]](List()).map { finalizers =>
+        UseToken(
+          acquire = self.setup.flatMap { tokenR =>
+            tokenR.acquire.flatMap { r =>
+              f0(r).setup.flatMap { tokenR1 =>
+                tokenR1.acquire.ensuring(finalizers.update(fs => tokenR1.release :: fs))
+              }
+            }.ensuring(finalizers.update(fs => tokenR.release :: fs))
+          },
+          release = finalizers.get.flatMap(fs => IO.foreach(fs.reverse)(identity)).uninterruptible
```

Comment:
Minor: you can use `collectAll` instead of `foreach` with `identity` (same below).

Reply:
Oh right thanks. These new names threw me off :-)